### PR TITLE
Provide option to disable calico-typha

### PR DESCRIPTION
**Pull request by vpnachev**
**What this PR does / why we need it**:
Provide option to disable calico-typha.
**Which issue(s) this PR fixes**:
Contributes to  https://github.com/gardener/gardener/issues/1579

**Special notes for your reviewer**:

**Release note**:

```improvement user
Calico Typha can be disabled for shoot clusters via `Shoot.spec.networking.providerConfig.typha.enabled=false`. Please read the [docs](https://docs.projectcalico.org/v3.9/reference/typha/) before to consider to disable typha.
```

### CODE REVIEW

**Comment by DockToFuture** [comment]
I'm wondering why we are deploying  ```calico-typha-horizontal-autoscaler```, ```calico-typha-vertical-autoscaler``` in to the cluster with ```replica=0``` when we disable typha. I would expect that we completely remove all deployments, svc, sa and so on, if we disable the addon.

**Comment by vpnachev** [comment]
@DockToFuture I have no strong opinion about what should be deployed and what not when typha is disabled. I just started with the very minimalistic approach. 
Basically, there is a lot of other resources like (Cluster)Role(Bindings), PodDisruptionBudget, and so on, that I do not want to touch at the first time.
I think to move all typha related resources to dedicated file(s) and they will not be rendered when typha is disabled - wdyt?

**Comment by rfranzke** [comment]
I agree with @DockToFuture, and I like your proposal @vpnachev.

**Comment by DockToFuture** [comment]
I like the idea of dedicated files.

**Comment by vpnachev** [comment]
@DockToFuture @zanetworker , this PR is ready for review.

File: controllers/networking-calico/charts/internal/calico/templates/rbac.yaml [hunk 1]
```diff
@@ -208,89 +210,3 @@ subjects:
 - kind: ServiceAccount
   name: calico-kube-controllers
   namespace: kube-system
----
-apiVersion: {{ include "rbacversion" . }}
```
Comment (rfranzke) [inline controllers/networking-calico/charts/internal/calico/templates/rbac.yaml]: Where did this cluster role go?

File: controllers/networking-calico/example/20-network.yaml [hunk 2]
```diff
@@ -50,5 +50,6 @@ spec:
     ipam:
       type: host-local
       cidr: usePodCIDR
-#    ipAutodetectionMethod: interface=eth0
-#    ipip: CrossSubnet
\ No newline at end of file
+#   ipAutodetectionMethod: interface=eth0
+#   ipip: CrossSubnet
+#   enableTypha: true
```
Comment (rfranzke) [inline controllers/networking-calico/example/20-network.yaml]:
```suggestion
#   typha:
#     enabled: true  
```

File: controllers/networking-calico/charts/internal/calico/templates/psp/calico-typha-clusterrole.yaml [hunk 3]
```diff
@@ -1,4 +1,4 @@
-
+{{- if .Values.config.enableTypha }}
```
[the remark this answers is not in the document]
Comment (vpnachev) [reply controllers/networking-calico/charts/internal/calico/templates/psp/calico-typha-clusterrole.yaml]: My idea was to move out `typha` related stuffs to a dedicated file only from `calico.yaml`.
Most of the RBAC resource were in separate files, so I just exclude them when typha is disabled.
I will merge the typha related RBAC files.

> Maybe it makes sense to even split typha/vertical-autoscaler/horizontal-autoscaler into three dedicated files.

I have no strong opinion, so do I have to do it or not?

File: controllers/networking-calico/charts/internal/calico/templates/rbac.yaml (hunk 1 above)
Comment (vpnachev) [reply controllers/networking-calico/charts/internal/calico/templates/rbac.yaml]: here https://github.com/gardener/gardener-extensions/pull/432/files#diff-c202b3b53bd963173804776737230878R28

File: controllers/networking-calico/charts/internal/calico/templates/psp/calico-typha-clusterrole.yaml (hunk 3 above)
[the remark this answers is not in the document]
Comment (rfranzke) [reply controllers/networking-calico/charts/internal/calico/templates/psp/calico-typha-clusterrole.yaml]: I think how it is now it's kind of inconsistent. Either put everything together in one file, or separate every resource into a dedicated file and not only RBAC. 

> Maybe it makes sense to even split typha/vertical-autoscaler/horizontal-autoscaler into three dedicated files.

No strong opinion either, I personally would probably do it, but then I would also separate the `calico-kube-controllers` related stuff into a separate file. Then we have four files for the four components. But as said, no strong opinion.

File: controllers/networking-calico/charts/internal/calico/templates/rbac.yaml (hunk 1 above)
Comment (rfranzke) [reply controllers/networking-calico/charts/internal/calico/templates/rbac.yaml]: Thanks!

File: controllers/networking-calico/charts/internal/calico/templates/psp/calico-typha-clusterrole.yaml (hunk 3 above)
[the remark this answers is not in the document]
Comment (vpnachev) [reply controllers/networking-calico/charts/internal/calico/templates/psp/calico-typha-clusterrole.yaml]: Now everything is merged into just couple of files:
1. `calico.yaml` - the minimal calico resource that are deployed always
1. `calico-typha.yaml` - resources that are deployed only when typha is enabled
1. Three files containing configmaps used for sha256 sums calculations.

File: controllers/networking-calico/charts/internal/calico/values.yaml [hunk 4]
```diff
@@ -4,7 +4,9 @@ config:
   backend: bird
   ipam:
     type: "host-local"
-#    subnet: "usePodCidr"
+#   subnet: "usePodCidr"
+  typha: {}
+#   enabled: true
```
Comment (rfranzke) [inline controllers/networking-calico/charts/internal/calico/values.yaml]: Can we uncomment this as this should be the default? I know, it's overwritten when actually rendering the chart, though, just to be safe.

Comment (vpnachev) [reply controllers/networking-calico/charts/internal/calico/values.yaml]: done

**Comment by ialidzhikov** [comment]
> What was the reason to put everything in 2 files? According to https://helm.sh/docs/topics/chart_best_practices/templates/:
> 
> > Each resource definition should be in its own template file.
> 
> Currently calico-typha.yaml is 495 lines long, calico.yaml - 861.

We can ignore this comment for now and fix it later.

**Comment by vpnachev** [comment]
Added end-user docs about how to disable Typha.
Thanks @ialidzhikov !

**Pull request merged**